### ISSUE TITLE
Weird retina support resizing inside Block Editor

### ISSUE THREAD

**Issue by maryfries** [jmoenig/Snap--Build-Your-Own-Blocks#1766]
![image](https://user-images.githubusercontent.com/6549787/27156697-46ca73b0-512c-11e7-81ce-5abee6bb6a34.png)

![image](https://user-images.githubusercontent.com/6549787/27156701-4b0e9bc2-512c-11e7-92ed-7f814cfd22c7.png)

**Comment by cycomachead** [comment]
Did you zoom blocks with the block editor open? I don't think this is a retina issue, just a really old bug.

**Comment by brianharvey** [comment]
Switching into or out of retina with a block editor open sometimes does strange things.  In a way I guess that counts as zooming.

**Comment by cycomachead** [comment]
> Switching into or out of retina with a block editor open sometimes does strange things. In a way I guess that counts as zooming.


oh, that also makes sense. Many things get broken with block editors open (like making/opening new projects, and changing the language.)

**Comment by brianharvey** [comment]
Should we rename this issue to "automatically close block editor before carrying out user instructions that break it"?

**Comment by cycomachead** [comment]
We don't need to close the block editor... We need to just call `.fixLayout()` or something (at least in this case). I'm pretty sure this is merely forgotten because the block editor is a child of the world and not the IDE (and the call is typically` ide.fixLayout()`). 

(Project switching is logged as a separate issue already, and that does make sense to close the block editor, of course.)

**Comment by maryfries** [comment]
@cycomachead "Did you zoom blocks with the block editor open?" Yes, I think so. Either that or I changed in and out of Retina support with it open. I forget. I took those screenshots a week or two ago and just ran across them again. ;)

**Comment by cycomachead** [comment]
Cool, thanks for checking in. It's fairly easy to reproduce in this case so that's good to know. 


-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by maryfries** [comment]
Sometimes, it gets _really_ weird.

![image](https://user-images.githubusercontent.com/6549787/27547101-de66b370-5a62-11e7-9b87-1f43cba3a1b8.png)

Note the OK, Apply, and Cancel buttons...

**Comment by jmoenig** [comment]
I'm thinking about taking out the zoom blocks function, now that we can simply zoom everything with retina support...

**Comment by maryfries** [comment]
I'm not sure what it would mean to zoom with retina support, but we often use the zoom blocks feature in development. Retina support creates a lot of image-sizing hassles, and I usually turn it off to make script pics.

**Comment by jmoenig** [comment]
Regardless of whether your monitor support retina you can use your browser's "view -> zoom" setting to make the blocks *and* the UI elements all bigger (or smaller). Then, if you reload Snap with this zoom setting (just press the reload button) our retina-support mechanism makes sure that everything becomes high-res. In many ways, this is much better than what zoom-blocks does, that's why I'm contemplating taking that feature out again.

**Comment by maryfries** [comment]
But would that allow for control over the script pic size?

**Comment by jmoenig** [comment]
yes, same way. In the meantime it's best to make sure all block editors are closed when you zoom blocks.

**Comment by maryfries** [comment]
OK, but users won't know that (and I'm, sadly, probably untrainable for something like this too). ;) 

I don't think it was zoom blocks that did it though; I think it was toggling retina support.

**Comment by brianharvey** [comment]
> I'm thinking about taking out the zoom blocks function

Noooooooo.....  Please no!  Unlike browser zoom, Snap<i>!</i> zoom blocks is remembered per user and applies only to Snap<i>!</i>.  It's a set-and-forget feature that is a great boon both to documentation-and-curriculum writers and to old people, both of which I am.  The only change you should make to zoom blocks is to make it work for the non-block elements of the make-a-block dialog.

**Comment by cycomachead** [comment]
NO  NO NO.

> yes, same way. In the meantime it's best to make sure all block editors are closed when you zoom blocks.

For one thing, browser zoom doesn't actually change the exported resolution of the script pics. The browser zoom feature just makes it such that every pixel in morphic is taken up by multiple pixels on screen. When you export something, it still appears at the native resolution. 

For another, this would suck for people with low vision. Zooming the entire UI and zooming blocks are two VERY different tools. One a "normal" system, you can control resolution and font size independently. Zoom blocks is controlling the font size, and using the browser zoom is controlling the (effective) resolution. 

And yes, there's the browser's min font size setting, and I do have this enabled and it's garbage. So many sites get messed up in weird ways, and even Snap<em>!</em> doesn't properly handle a scaled font because everything is off centered. 

Zooming all UI elements is not "better". It's different. We interact with text and UI differently. Text is a highly detailed mechanism that requires paying very close attention to read. UI elements (in general) are larger, colored, and have shapes and various other affordances explain their meanings. There's no right answer, but people with low version can work with comparatively normal sized UIs where they would require larger fonts. Not everything deserves equal scaling.

**Comment by cycomachead** [comment]
Oh, yeah, I spent a half hour or so looking into why this happens. 

For the life of me, I can't figure out why the block editor won't redraw itself... I've tried all sorts of `drawNew()` and `fixLayout()` calls.
